### PR TITLE
Adding CIAM E2E tests and improvements to build infrastructure

### PR DESCRIPTION
**Pull request by bgavrilMS**
# Adding CIAM E2E tests and improvements to build infrastructure

- [x] You've read the [Contributor Guide](https://github.com/AzureAD/microsoft-identity-web/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/AzureAD/microsoft-identity-web/blob/master/CODE_OF_CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [ ] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Summary of the changes (Less than 80 chars)

## Description

CIAM tests

- Webapp calls web api local

- Web app calls web api (graph) disabled on the ADO for now. will be revisited(The entire file is disabled, not just this test)

Improvements to the YAML files

- ID Web now has separate tasks for Building and running tests (One for unit tests, integration tests and E2E tests)

- Separating these tasks should make it easier to onboard to Onebranch if we run into issues getting E2E tests running on the OneBranch agents.

- ID Web can now leverage the test rerun feature in the VSTest ADO task to rerun failed tests. This will help improve overall test reliability.

- Added logic to enable E2E tests to attempt a restart process restart a few times when it crashes prematurely. This should further improve reliability for the E2E tests
- Screenshots are now captured when E2E Ui tests fail and are uploaded to the build run.

Fixes #{bug number} (in this specific format)

Continuation of https://github.com/AzureAD/microsoft-identity-web/pull/2728

### CODE REVIEW

File: build/template-run-unit-tests.yaml
Comment (bgavrilMS) [inline build/template-run-unit-tests.yaml:13]: This is great!

File: build/template-run-unit-tests.yaml
Comment (bgavrilMS) [inline build/template-run-unit-tests.yaml:18]: Hmm, but you have to figure out a way to break the build if the integration tests fail.

File: tests/DevApps/ciam/myWebApi/Properties/launchSettings.json
Comment (bgavrilMS) [inline tests/DevApps/ciam/myWebApi/Properties/launchSettings.json:12]: Should it ever use the `https` profile?

File: build/template-restore-build-MSIdentityWeb.yaml
Comment (bgavrilMS) [inline build/template-restore-build-MSIdentityWeb.yaml:22]: I assume this is temporary ? "build restore template" is not for running tests.

File: build/template-run-unit-tests.yaml
Comment (trwalke) [reply build/template-run-unit-tests.yaml]: This was for testing. removed

File: tests/DevApps/WebAppCallsMicrosoftGraph/Startup.cs
Comment (bgavrilMS) [inline tests/DevApps/WebAppCallsMicrosoftGraph/Startup.cs:6]: Are these changes needed? You don't even run this app?

File: tests/E2E Tests/WebAppUiTests/UiTestHelpers.cs
Comment (bgavrilMS) [inline tests/E2E Tests/WebAppUiTests/UiTestHelpers.cs:332]: for i=1;i<=3;i++ ?

File: tests/E2E Tests/WebAppUiTests/UiTestHelpers.cs
Comment (bgavrilMS) [inline tests/E2E Tests/WebAppUiTests/UiTestHelpers.cs:328]: nit: await Task.Delay(5000)

File: tests/E2E Tests/WebAppUiTests/UiTestHelpers.cs
Comment (bgavrilMS) [inline tests/E2E Tests/WebAppUiTests/UiTestHelpers.cs:356]: There is a "Process.IsAlive"

**Comment by bgavrilMS** [comment]
I can't approve it because I opened the PR. Signing off.

File: build/template-restore-build-MSIdentityWeb.yaml
Comment (trwalke) [reply build/template-restore-build-MSIdentityWeb.yaml]: Yes. this was removed from here

File: tests/DevApps/WebAppCallsMicrosoftGraph/Startup.cs
Comment (trwalke) [reply tests/DevApps/WebAppCallsMicrosoftGraph/Startup.cs:6]: This was a workaround I attempted to get the second test functioning on the ADO pipeline. Didnt work so i will remove

File: tests/DevApps/ciam/myWebApi/Properties/launchSettings.json
Comment (trwalke) [reply tests/DevApps/ciam/myWebApi/Properties/launchSettings.json:12]: Not sure, maybe for testing some specific scenarios. I would leave it there

File: tests/E2E Tests/WebAppUiTests/UiTestHelpers.cs
Comment (trwalke) [reply tests/E2E Tests/WebAppUiTests/UiTestHelpers.cs:328]: I am trying to avoid making these methods async because they prevent the use of the out parameter. I guess I can return the result as a Tuple

File: tests/E2E Tests/WebAppUiTests/UiTestHelpers.cs
Comment (trwalke) [reply tests/E2E Tests/WebAppUiTests/UiTestHelpers.cs:356]: Yes, will be using that

**Comment by trwalke** [comment]
> I can't approve it because I opened the PR. Signing off.

I can approve for you :). Still need a second approval

**Review by trwalke — APPROVED** [review]

**Review by jmprieur — CHANGES_REQUESTED** [review]
Please address piushing tests and coveragev to ado as before

File: build/template-restore-build-MSIdentityWeb.yaml
Comment (trwalke) [reply build/template-restore-build-MSIdentityWeb.yaml:17]: OK, I will take a look

Comment (jmprieur) [reply build/template-restore-build-MSIdentityWeb.yaml:17]: I see that now the tests are pushed to Azure DevOps, but the coverage has all the assemblies, not only the ones we care about (we don't care about the coverage of test apps, etc ..)

See:
-  https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=1285178&view=codecoverage-tab with your PR, against
- https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=1285100&view=codecoverage-tab before your PR

(coverage only for the product assemblies)

Comment (trwalke) [reply build/template-restore-build-MSIdentityWeb.yaml:17]: @jmprieur latest changes have the correct code coverage
https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=1285575&view=codecoverage-tab

File: tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs
Comment (trwalke) [reply tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs:48]: Graph and flow are 2 different words so I used `GraphFlow` based on pascal casing rules

File: tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs
Comment (trwalke) [reply tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs:211]: Yes, this is the type returned by `IBrowserContext.NewPageAsync()`. Using it here lets me have access to it in the full scope of the test method

File: build/template-run-unit-tests.yaml
Comment (jmprieur) [reply build/template-run-unit-tests.yaml:25]: out of curiosity why is it great?

**Review by jmprieur — APPROVED** [review]
LGTM
Thanks @trwalke 

Once merged, you will need to cherry-pick to the rel/v2 branch too.

File: build/template-run-unit-tests.yaml
Comment (trwalke) [reply build/template-run-unit-tests.yaml:25]: Because it allows integration tests to rerun with 3 attempts if the test is flaky. Most of the time, the flaky tests we have in MSAL are worked around using this feature, especially for tests where it is challenging to find the root cause of the flakiness. These tests will show up in the "Passed on rerun" category in the build results. Having a test rerun itself and passing is better than the build failing entirely and needing to be rerun manually.

image

**Pull request merged**
